Update visual style of CupertinoSwitch to match iOS 13

brandondiamond opened the pull request:
## Description

Updates the CupertinoSwitch widget to reflect the revised design of switches in iOS 13. In particular, the border no longer animates in response to user interaction. Instead, the background fades between colors.

**After:**
https://i.imgur.com/dMV7T5G.gif (slow 5x)
https://i.imgur.com/arKDoNk.gif (normal)

**Before:**
https://i.imgur.com/SxDYis4.gif (slow 5x)
https://i.imgur.com/k7HCNJU.gif (normal)

## Related Issues

Issue #33797

## Tests

Added a test to verify golden images of the slider in on and off states.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

brandondiamond committed:
Update visual style of cupertino switch to match ios13

brandondiamond committed:
Remove math import, as it's no longer necessary.

LongCatIsLooong commented:
Hi @brandondiamond thank you for contributing, the PR looks awesome! However tests are mandatory for PRs like this. Could you add tests to ensure it works as expected? 

In case you want to add golden tests: https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package:flutter

brandondiamond commented:
Thanks so much for the quick review! I'll get started on those golden tests right away. Given that this is a visual change, are there any additional tests (beyond the golden tests) that you would suggest?

brandondiamond commented:
Added GIFs of before/after:

**After:**
https://i.imgur.com/dMV7T5G.gif (slow 5x)
https://i.imgur.com/arKDoNk.gif (normal)

**Before:**
https://i.imgur.com/SxDYis4.gif (slow 5x)
https://i.imgur.com/k7HCNJU.gif (normal)

LongCatIsLooong commented:
@brandondiamond thanks for the GIFs!

Now that the track color change is animated, another thing we could add a test for is to verify the track color doesn't just jump between `lightBackgroundGray` and `widget.activeColor`. But in my opinion adding golden images for both active/inactive switches should be good enough.

justinmc reviewed (commented):
Thanks for the PR, this is great!

+1 for tests.  I think that one golden image for on and one for off should be fine.

One very small visual nitpick.  I think that our shadow extends below the switch track when it shouldn't anymore.

|              | This Branch  | iOS 13          |
| ---------- | ----------------- | ----------------- |
| On        | <img width="250" src="https://user-images.githubusercontent.com/389558/61395510-7261cf80-a87a-11e9-819d-1daf48b5c7ac.png" /> | <img width="250" alt="Screen Shot 2019-07-17 at 9 52 57 AM" src="https://user-images.githubusercontent.com/389558/61395340-126b2900-a87a-11e9-9fcb-829ea05406bd.png"> |
| Off        | <img width="250" src="https://user-images.githubusercontent.com/389558/61395511-7261cf80-a87a-11e9-9bff-634d962e369d.png" /> | <img width="250" alt="Screen Shot 2019-07-17 at 9 52 49 AM" src="https://user-images.githubusercontent.com/389558/61395338-1008cf00-a87a-11e9-8351-1be826650aea.png"> |

brandondiamond committed:
Add clip to prevent thumb from exceeding track.

brandondiamond commented:
Thanks Justin! I added a clip to ensure the thumb shadow doesn't exceed the track. I know clipping can be expensive so wanted to verify that it'd be okay in this case.

I also wanted to be mindful of antialiasing (since the clip is curved), so I used saveLayer to group the thumb drawing commands together. There's an additional save() and restore() around the clip because I wasn't certain whether there's an automatic save/restore around each paint() -- particularly for render objects that don't enforce repaint boundaries; I don't want subsequent drawing operations to be affected by the clip.

I'm very grateful for the opportunity to contribute -- thanks again for reviewing :)

brandondiamond committed:
Add golden tests for restyled switch

brandondiamond commented:
Added golden tests and generated goldens. I followed the instructions [here](https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package:flutter) but ran into permission issues when attempting to push my golden assets commit to the repo.

What's the correct way to push a change to the goldens repo without special permissions?

justinmc commented:
Thanks!  I tried running this again and now that it's clipped it looks perfect to my eye.

Regarding performance and antialiasing, I'm going to track down someone that can confirm.

Regarding permissions on the goldens repo, I believe you'll have to fork that repo and open a PR like you did with this PR.  I know we tightened up this process recently so I could be wrong, but that's how I did it a few months ago.  If you post back with a link to the goldens PR, I can review it really quickly though.

brandondiamond commented:
Awesome, thanks Justin :)

Here's the goldens PR:
https://github.com/flutter/goldens/pull/40

Once that goes through, I'll update this PR to reference the commit that includes the new goldens.

LongCatIsLooong commented on packages/flutter/test/cupertino/switch_test.dart:
```diff
@@ -513,4 +513,54 @@ void main() {
+            return Center(
+              child: RepaintBoundary(
+                  child: CupertinoSwitch(
+                  key: switchKey,
```
nit: `key` should be further indented.

LongCatIsLooong commented on packages/flutter/test/cupertino/switch_test.dart:
```diff
@@ -513,4 +513,54 @@ void main() {
+          builder: (BuildContext context, StateSetter setState) {
+            return Center(
+              child: RepaintBoundary(
+                  child: CupertinoSwitch(
```
nit: too much indentation.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -507,12 +502,17 @@ class _RenderCupertinoSwitch extends RenderConstrainedBox {
     );
     final double thumbCenterY = offset.dy + size.height / 2.0;
 
+    canvas.save();
```
Most of the time we just use `context.pushClipRRect`. An example:
https://github.com/flutter/flutter/blob/ab26e1b4a7cdefb5b9b2ba53e35c68ec3116f0e3/packages/flutter/lib/src/rendering/proxy_box.dart#L1371

Relaying @liyuqian 's answer, if it is just the shadow you want to clip off, it's probably not worth doing `saveLayer` (as there's no overlapping draw calls at the edge of the `RRect`, if I understand it correctly).

LongCatIsLooong commented on packages/flutter/test/cupertino/switch_test.dart:
```diff
@@ -513,4 +513,54 @@ void main() {
+      ),
+      skip: !isLinux,
+    );
+
```
Could you also verify that 100 milliseconds (which is half of the full duration of the animation. Or whatever timing you prefer) into the animation, the switch still has the correct appearance? Something like this:
```
    await tester.tap(find.byKey(switchKey));
   // Kick off animation.
    await tester.pump()
    await tester.pump(const Duration(milliseconds: 100));
    await expectLater(
      find.byKey(switchKey),
      matchesGoldenFile(
        'switch.tap.turningOn.png',
        version: 0,
      ),
      skip: !isLinux,
    );
```

brandondiamond replied:
This is super cool. Thanks for sharing that snippet!

brandondiamond committed:
Add intermediate test and use different clipping mechanism.

brandondiamond commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -501,18 +501,16 @@ class _RenderCupertinoSwitch extends RenderConstrainedBox {
-    canvas.restore();
+    );
+
+    context.pushClipRRect(needsCompositing, Offset.zero, thumbBounds, trackRRect, (innerContext, _) {
```
I wasn't sure whether to shorten this line and I'm hesitant to run dartfmt on library source (though if that's standard practice, I'll do that and re-push)

LongCatIsLooong replied:
The analyzer isn't happy about this line. See https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#naming-rules-for-typedefs-and-function-variables
> If you have a callback with arguments but you want to ignore the arguments, give the type and names of the arguments anyway. That way, if someone copies and pastes your code, they will not have to look up what the arguments are.


```suggestion
    context.pushClipRRect(needsCompositing, Offset.zero, thumbBounds, trackRRect, (PaintContext innerContext, Offset offset) {
```

LongCatIsLooong replied:
Regarding dartfmt: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#formatting

> These guidelines have no technical effect, but they are still important purely for consistency and readability reasons.
> We do not yet use dartfmt. Flutter code tends to use patterns that the standard Dart formatter does not handle well. We are working with Dart team to make dartfmt aware of these patterns.

brandondiamond commented:
Updated the tests, generated new goldens (ready to go in https://github.com/flutter/goldens/pull/40 -- I'll update the golden hash here once that is merged), and switched to using the PaintingContext to push a clip.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -506,13 +501,16 @@ class _RenderCupertinoSwitch extends RenderConstrainedBox {
     final double thumbCenterY = offset.dy + size.height / 2.0;
-
-    _thumbPainter.paint(canvas, Rect.fromLTRB(
+    final thumbBounds = Rect.fromLTRB(
```
The analyzer is complaining about the missing type annotation.
```suggestion
    final Rect thumbBounds = Rect.fromLTRB(
```

brandondiamond replied:
Whoops, fixed!

brandondiamond committed:
Add missing type annotation

brandondiamond committed:
Add types and names to paint closure.

Co-Authored-By: LongCatIsLooong <31859944+LongCatIsLooong@users.noreply.github.com>

brandondiamond commented:
Thanks for the review! I'm going to read the style guide today to make sure I don't repeat these formatting mistakes [too often :)]

brandondiamond committed:
Bump goldens version to include new assets

brandondiamond committed:
Merge branch 'cupertino_switch' of github.com:brandondiamond/flutter into cupertino_switch

brandondiamond commented:
Goldens merged and bumped the golden version! Should be good to go.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -506,13 +501,16 @@ class _RenderCupertinoSwitch extends RenderConstrainedBox {
-    ));
+    );
+
+    context.pushClipRRect(needsCompositing, Offset.zero, thumbBounds, trackRRect, (PaintContext innerContext, Offset offset) {
```
`PaintContext` => `PaintingContext`, sorry for the wrong suggestion.

LongCatIsLooong replied:
```suggestion
    context.pushClipRRect(needsCompositing, Offset.zero, thumbBounds, trackRRect, (PaintingContext innerContext, Offset offset) {
```

brandondiamond committed:
PaintContext => PainterContext (whoops!)

brandondiamond commented:
No worries! I've made that mistake many times

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -506,13 +501,16 @@ class _RenderCupertinoSwitch extends RenderConstrainedBox {
-    ));
+    );
+
+    context.pushClipRRect(needsCompositing, Offset.zero, thumbBounds, trackRRect, (PainterContext innerContext, Offset offset) {
```
```suggestion
    context.pushClipRRect(needsCompositing, Offset.zero, thumbBounds, trackRRect, (PaintingContext innerContext, Offset offset) {
```

brandondiamond replied:
Aaaaand that's why I shouldn't quickly fix things on my phone. Sorry.

brandondiamond committed:
Fix typo (let's pretend this didn't happen)

Co-Authored-By: LongCatIsLooong <31859944+LongCatIsLooong@users.noreply.github.com>

LongCatIsLooong commented on packages/flutter/test/cupertino/switch_test.dart:
```diff
@@ -513,4 +513,66 @@ void main() {
+        version: 0,
+      ),
+      skip: !isLinux,
+    );  
```
> ⏩ RUNNING: cd .; grep --line-number --extended-regexp [[:blank:]]$ packages/flutter/lib/src/cupertino/switch.dart packages/flutter/test/cupertino/switch_test.dart
packages/flutter/test/cupertino/switch_test.dart:565:    );  
🕐 ELAPSED TIME: 0:00:00.019169 for grep --line-number --extended-regexp [[:blank:]]$ 
packages/flutter/lib/src/cupertino/switch.dart packages/flutter/test/cupertino/switch_test.dart in .: 
Whitespace detected at the end of source code lines.
Please remove:

Looks like there're trailing white spaces at the end of this line. Could you remove it?

```suggestion
    );
```

brandondiamond committed:
Remove trailing space

Co-Authored-By: LongCatIsLooong <31859944+LongCatIsLooong@users.noreply.github.com>

Piinks commented:
Currently blocked by #36690, which will resolve the golden sync issues that are causing the tests to fail here.

brandondiamond commented:
Thanks for looking into this. Please let me know if there is anything I can do to help.

Piinks reviewed (commented):
#36690 has landed. You may need to update your branch by pulling down the changes from master.

brandondiamond commented:
@Piinks Thanks! What's the best way to do this without messing up commit history?

Piinks commented:
You should be able to do a 
`git fetch upstream` 
and then 
`git pull upstream master`
This will create a commit that merges updates from master into your branch, which you can then push here. :)

brandondiamond committed:
Merge branch 'master' of github.com:flutter/flutter into cupertino_switch

brandondiamond commented:
Thanks so much! Just pushed -- hopefully all looks okay :)

Piinks reviewed (approved):
LGTM! Thank you for the contribution! 🎉

justinmc reviewed (approved):
LGTM.  Thanks for this!  I'll merge soon.
